DQMServices/Core/interface/ConcurrentMonitorElement.h

```diff
@@ -0,0 +1,140 @@
+#ifndef DQMServices_Core_ConcurrentMonitorElement_h
+#define DQMServices_Core_ConcurrentMonitorElement_h
+
+/* Encapsulate of MonitorElement to expose *limited* support for concurrency.
+ *
+ * ...
+ */
+
+#include <mutex>
+
+#include "DQMServices/Core/interface/MonitorElement.h"
+
+class ConcurrentMonitorElement
+{
+private:
+  mutable MonitorElement* me_;
+  mutable std::mutex lock_;
+
+public:
+  ConcurrentMonitorElement(void) :
+    me_(nullptr)
+  { }
+
+  explicit ConcurrentMonitorElement(MonitorElement* me) :
+    me_(me)
+  { }
+
+  // non-copiable
+  ConcurrentMonitorElement(ConcurrentMonitorElement const&) = delete;
+
+  // movable
+  ConcurrentMonitorElement(ConcurrentMonitorElement && other)
+  {
+    std::lock_guard<std::mutex>(other.lock_);
+    me_ = other.me_;
+    other.me_ = nullptr;
+  }
+
+  // not copy-assignable
+  ConcurrentMonitorElement& operator=(ConcurrentMonitorElement const&) = delete;
+
+  // move-assignable
+  ConcurrentMonitorElement& operator=(ConcurrentMonitorElement && other)
+  {
+    // FIXME replace with std::scoped_lock once C++17 is available
+    std::lock(lock_, other.lock_);
+    std::lock_guard<std::mutex> ours(lock_, std::adopt_lock);
+    std::lock_guard<std::mutex> others(other.lock_, std::adopt_lock);
+    me_ = other.me_;
+    other.me_ = nullptr;
+    return *this;
+  }
+
+  // nothing to do, we do not own the MonitorElement
+  ~ConcurrentMonitorElement(void) = default;
+
+  // expose as a const method to mean that it is concurrent-safe
+  template <typename... Args>
+  void fill(Args && ... args) const
+  {
+    std::lock_guard<std::mutex> guard(lock_);
+    me_->Fill(std::forward<Args>(args)...);
+  }
+
+  // expose as a const method to mean that it is concurrent-safe
+  void shiftFillLast(double y, double ye = 0., int32_t xscale = 1) const
+  {
+    std::lock_guard<std::mutex> guard(lock_);
+    me_->ShiftFillLast(y, ye, xscale);
+  }
+
+  // reset the internal pointer
+  void reset()
+  {
+    std::lock_guard<std::mutex> guard(lock_);
+    me_ = nullptr;
+  }
+
+  operator bool() const
+  {
+    std::lock_guard<std::mutex> guard(lock_);
+    return (me_ != nullptr);
+  }
+
+  // non-const methods to manipulate axes and titles.
+  // these are not concurrent-safe, and should be used only when the underlying
+  // MonitorElement is being booked.
+  void setTitle(std::string const& title)
+  {
+    me_->setTitle(title);
```

Comment:
hi @fwyzard - ignorant question - why not add the same lock_guards in these functions (or would that not be sufficient for making them concurrent safe?)

Reply:
I think that you are right, and it would indeed make these methods concurrent safe.

The reason I did not mark them as such is the DQM usage pattern, or at least the one I have in mind:
  - the `ConcurrentMonitorElements` (CMEs) are booked in the (global) begin run transition, where they are accessible as non-`const` objects; 
  - the CMEs are later filled in the event loop, where they are accessible only as `const` objects to emphasise that they are concurrent-safe.

I can add the locks also for the operations on the titles and labels; however, I expect those to be set when the CMEs are created and available as non-`const`, are used only by one thread, and anyway already serialised by the big DQM lock.

An alternative approach I've been playing with - which is more flexible but slightly less clean - is to let the users create the histograms on their own via `auto h= std::make_unique<TH1F>(name, ...)` etc., customise the axes, titles, labels etc via access the full ROOT object interface, and when done transfer ownership to the DQMStore via a call to `book1D(name, std::move(h))`.
This should be just as efficient from the memory point of view (no `Clone()` involved), allow more flexibility, at the cost of one more line of code (per histogram) and with the small risk that the users may hold on to the initial pointer and use it later on to bypass the lock.

Reply:
Right - I agree its the best usage pattern to have all of these CMEs beautified in the begin run transitions - but if the code allows it elsewhere someone will do... 

Your thoughts about transferring ownership of a TH1F are easier in the sense that there are a lot of functions that would need wrapping if the direct access to the TH* is removed (a good thing) - but the tradeoff is the ease of any future underlying histogram interface changes.

The second thought is beyond this PR - do you want to add locks for now?

Reply:
I think I'd rather start with the present interface, and update it as needed when we start transitioning the analyzers to use it - we may encounter some patterns I haven't anticipated yet...